tests/backends/postgresql/test_server_side_cursors.py

```diff
@@ -23,8 +23,13 @@ def setUpTestData(cls):
 
     def inspect_cursors(self):
         with connection.cursor() as cursor:
+            # Note: Postgres shows a cursor with no name if prepared statements
+            # have been used. Avoid to fetch it or this very query below will
+            # be reported as created by a named cursor.
             cursor.execute(
-                "SELECT {fields} FROM pg_cursors;".format(fields=self.cursor_fields)
+                "SELECT {fields} FROM pg_cursors where name <> ''".format(
```

Comment:
What happened here, is this a behavior change with psycopg3 and cursor naming?

Reply:
I have a gut feeling that something is leaking here, e.g. cursor is not properly closed or is being shared when it shouldn't be. When running tests in random order
```
./runtests.py backends.postgresql.test_server_side_cursors --shuffle
```
you can notice that only one test fails (different each time).

Reply:
This could be completely unrelated, but I'm also trying to find the cause of some new isolation issue on CockroachDB when running with `model_options.test_default_related_name model_options.test_tablespaces` (running only the latter file works fine). For some reason, when preceded by at least one test from `test_default_related_name`, `connection` is `None` for the first tablespaces test, i.e.
psycopg3
```
======================================================================
ERROR: test_tablespace_for_indexed_field (model_options.test_tablespaces.TablespacesTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/tim/code/django/django/test/testcases.py", line 1604, in skip_wrapper
    return test_func(*args, **kwargs)
  File "/home/tim/code/django/tests/model_options/test_tablespaces.py", line 78, in test_tablespace_for_indexed_field
    sql = sql_for_table(Article).lower()
  File "/home/tim/code/django/tests/model_options/test_tablespaces.py", line 18, in sql_for_table
    editor.create_model(model)
  File "/home/tim/code/django/django/db/backends/base/schema.py", line 164, in __exit__
    self.execute(sql)
  File "/home/tim/code/django/django/db/backends/postgresql/schema.py", line 51, in execute
    sql = compose(sql, params, self.connection.connection)
  File "/home/tim/code/django/django/db/backends/postgresql/operations.py", line 394, in compose
    return ClientCursor(connection).mogrify(str(query), params)
  File "/home/tim/.virtualenvs/django39/lib/python3.9/site-packages/psycopg/cursor.py", line 674, in __init__
    super().__init__(connection)
  File "/home/tim/.virtualenvs/django39/lib/python3.9/site-packages/psycopg/cursor.py", line 65, in __init__
    self._pgconn = connection.pgconn
AttributeError: 'NoneType' object has no attribute 'pgconn'
```
psycopg2:
```
======================================================================
ERROR: test_tablespace_for_indexed_field (model_options.test_tablespaces.TablespacesTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/tim/code/django/django/test/testcases.py", line 1604, in skip_wrapper
    return test_func(*args, **kwargs)
  File "/home/tim/code/django/tests/model_options/test_tablespaces.py", line 78, in test_tablespace_for_indexed_field
    sql = sql_for_table(Article).lower()
  File "/home/tim/code/django/tests/model_options/test_tablespaces.py", line 17, in sql_for_table
    with connection.schema_editor(collect_sql=True) as editor:
  File "/home/tim/code/django/django/db/backends/base/schema.py", line 164, in __exit__
    self.execute(sql)
  File "/home/tim/code/django/django/db/backends/postgresql/schema.py", line 51, in execute
    sql = compose(sql, params, self.connection.connection)
  File "/home/tim/code/django/django/db/backends/postgresql/operations.py", line 399, in compose
    return connection.cursor().mogrify(str(query), params).decode()
AttributeError: 'NoneType' object has no attribute 'cursor'
```

Reply:
Not sure why this would happen, but can you add a print of the stack in `django.db.backends.base.base.BaseDabaseWrapper.close`. I am wondering if the first test managed to close the connection, or if we simply have none yet.

Reply:
Interesting. I have enable postgres statement logging and when the server side cursor tests fail the logs show:
```
2022-12-03 12:44:10.574 CET [12212] LOG:  execute _pg3_2: SELECT name, statement, is_holdable, is_binary, is_scrollable, creation_time FROM pg_cursors
```
and when they work they show:
```
2022-12-03 12:44:10.570 CET [12212] LOG:  statement: SELECT name, statement, is_holdable, is_binary, is_scrollable, creation_time FROM pg_cursors
```

Why do we have `statement` once and `execute` in the other case?

Reply:
Ah lol, `_pg3` is used by psycopg3. It seems to prepare statements by default after a few executions. 

This fixes it:
```diff
diff --git a/django/db/backends/postgresql/base.py b/django/db/backends/postgresql/base.py
index 6b9b145fe2..5c467b9c0a 100644
--- a/django/db/backends/postgresql/base.py
+++ b/django/db/backends/postgresql/base.py
@@ -259,7 +259,9 @@ class DatabaseWrapper(BaseDatabaseWrapper):
 
         def _get_new_connection(self, conn_params):
             ctx = get_adapters_template(settings.USE_TZ, self.timezone)
-            return self.Database.connect(**conn_params, context=ctx)
+            conn = self.Database.connect(**conn_params, context=ctx)
+            conn.prepare_threshold = None
+            return conn
 
         def _configure_cursor(self, cursor):
             # Register the cursor timezone only if the connection disagrees, so that
diff --git a/tests/backends/postgresql/test_server_side_cursors.py b/tests/backends/postgresql/test_server_side_cursors.py
index af2bf9b1bb..9e4b19d721 100644
--- a/tests/backends/postgresql/test_server_side_cursors.py
+++ b/tests/backends/postgresql/test_server_side_cursors.py
@@ -23,13 +23,8 @@ class ServerSideCursorsPostgres(TestCase):
 
     def inspect_cursors(self):
         with connection.cursor() as cursor:
-            # Note: Postgres shows a cursor with no name if prepared statements
-            # have been used. Avoid to fetch it or this very query below will
-            # be reported as created by a named cursor.
             cursor.execute(
-                "SELECT {fields} FROM pg_cursors where name <> ''".format(
-                    fields=self.cursor_fields
-                )
+                "SELECT {fields} FROM pg_cursors".format(fields=self.cursor_fields)
             )
             cursors = cursor.fetchall()
         return [self.PostgresCursor._make(cursor) for cursor in cursors]
```
I wonder if we want to use prepared queries?

Reply:
> I wonder if we want to use prepared queries?

Prepared statements [don't work with external connection poolers](https://www.psycopg.org/psycopg3/docs/advanced/prepare.html), such as `PgBouncer`, so we should disable them (IMO).

Reply:
Also, we can pass `prepare_threshold=None` to the `connect()`.

Reply:
Those are good points. So +1 for disabling it by default. I also think it might make the tests faster because we usually use savepoints + rollback. So those queries that are actually getting prepared are only executed once and then deallocated again :D

Reply:
Completely unscientific but with `--parallel=8` the tests are indeed consistently a bit faster (3 runs each) by roughly 3-4 seconds (247 seconds vs 251 seconds) when prepared statements are disabled.

Reply:
Regarding the failure of `test_tablespace_for_indexed_field (model_options.test_tablespaces.TablespacesTests)` on CockroachDB described a few comments above, the difference on CockroachDB is `DatabaseFeatures.can_rollback_ddl = False` so `SchemaEditor.atomic_migration = False`. The issue doesn't happen on PostgreSQL where atomic migrations are used because a side effect of atomic migrations is opening a connection:
```
  File "/home/tim/code/django/django/db/backends/base/schema.py", line 159, in __enter__
    self.atomic.__enter__()
  File "/home/tim/code/django/django/db/transaction.py", line 198, in __enter__
    if not connection.get_autocommit():
  File "/home/tim/code/django/django/db/backends/base/base.py", line 466, in get_autocommit
    self.ensure_connection()
  File "/home/tim/code/django/django/utils/asyncio.py", line 26, in inner
    return func(*args, **kwargs)
  File "/home/tim/code/django/django/db/backends/base/base.py", line 291, in ensure_connection
    self.connect()
```
The following addition to `BaseDatabaseSchemaEditor.__enter__()` solves the issue, but may be unnecessary overhead for other backends. It could be fixed with an override in django-cockroachdb's `SchemaEditor` instead.
```diff
diff --git a/django/db/backends/base/schema.py b/django/db/backends/base/schema.py
index fe31967ce2..ba29179c7c 100644
--- a/django/db/backends/base/schema.py
+++ b/django/db/backends/base/schema.py
@@ -156,6 +156,8 @@ class BaseDatabaseSchemaEditor:
         if self.atomic_migration:
             self.atomic = atomic(self.connection.alias)
             self.atomic.__enter__()
+        else:
+            self.connection.ensure_connection()
         return self
 
     def __exit__(self, exc_type, exc_value, traceback):
```